feat(`server` + `client`): adding a streaming option for batched requests (#4347)
trpc/trpc: www/docs/client/links/httpBatchStreamLink.md

Reviewer: Some note about browser compatibility or something here?

Is there something we can do to automatically degrade when we detect that `fetch`-polyfill or the browser doesn't support it?

Author: I did look for the fetch stuff, couldn't find a good way to detect "the shape of `response.body`". But having a `response.body` return a stream is *almost* as old as having support for `fetch` itself (Safari 11 vs Safari 10), and in addition to that, my implementation also supports undici and node-fetch. So I think at that point it's on the devs to decide whether they want a modern streaming request or not.

---
id: httpBatchStreamLink
title: HTTP Batch Stream Link
sidebar_label: HTTP Batch Stream Link
slug: /links/httpBatchStreamLink
---

> `unstable_httpBatchStreamLink` (experimental) is a [**terminating link**](./index.md#the-terminating-link) that batches an array of individual tRPC operations into a single HTTP request that's sent to a single tRPC procedure (equivalent to [`httpBatchLink`](./httpBatchLink.md)), but doesn't wait for all the responses of the batch to be ready and streams the responses as soon as any data is available.

Author: ```suggestion
`unstable_httpBatchStreamLink` (unstable) is a [**terminating link**](./index.md#the-terminating-link) that batches an array of individual tRPC operations into a single HTTP request that's sent to a single tRPC procedure (equivalent to [`httpBatchLink`](./httpBatchLink.md)), but doesn't wait for all the responses of the batch to be ready and streams the responses as soon as any data is available.
```

## Usage

All usage and options are identical to [`httpBatchLink`](./httpBatchLink.md).

## Streaming mode (experimental)

> > ⚠️ This link is experimental and may change in the future.

Reviewer: Use the term `unstable` instead - https://trpc.io/docs/faq#unstable

The way I reason about things is that "experimental" is badly tested things that might be changed or dropped and unstable things are things that we sort-of committed to but might have issues

tl;dr: experimental ≈ alpha, unstable ≈ beta

I think this should be 'beta"

> ## Streaming mode (experimental)
> 
> > ⚠️ This link is experimental and may change in the future.

Author: ```suggestion
## Streaming mode (unstable)

> ⚠️ This link is unstable and may change in the future.
```

:::info
We have prefixed this as `unstable_` as it's a new API, but you're safe to use it! [Read more](/docs/faq#unstable).
:::

## Usage

> All usage and options are identical to [`httpBatchLink`](./httpBatchLink.md).

You can import and add the `httpBatchStreamLink` to the `links` array as such:

```ts title="client/index.ts"
import { createTRPCProxyClient, httpBatchStreamLink } from '@trpc/client';
import type { AppRouter } from '../server';

const client = createTRPCProxyClient<AppRouter>({
  links: [
    httpBatchStreamLink({
      url: 'http://localhost:3000',
    }),
  ],
});
```

After that, you can make use of batching by setting all your procedures in a `Promise.all`. The code below will produce exactly **one** HTTP request and on the server exactly **one** database query:

```ts
const somePosts = await Promise.all([
  trpc.post.byId.query(1),
  trpc.post.byId.query(2),
  trpc.post.byId.query(3),
]);
```

## Streaming mode

> ⚠️ This link is unstable and may change in the future.

> When batching requests together, the behavior of a regular `httpBatchStreamLink` is to wait for all requests to finish before sending the response. If you want to send responses as soon as they are ready, you can use this link. This is useful for long-running requests.

Author: This sentence is a description of the differences between `httpBatchStreamLink` and `httpBatchLink`. It is normal to have `httpBatchLink` mentioned in it.

```suggestion
When batching requests together, the behavior of a regular `httpBatchLink` is to wait for all requests to finish before sending the response. If you want to send responses as soon as they are ready, you can use `httpBatchStreamLink` instead. This is useful for long-running requests.
```

      url: 'http://localhost:3000',
    }),
  ],
});
```

Compared to a regular `httpBatchLink`, a `unstable_httpBatchStreamLink` will:

- Cause the requests to be sent with a `Trpc-Batch-Mode: stream` header
- Cause the response to be sent with a `Transfer-Encoding: chunked` and `Vary: trpc-batch-mode` headers
- Remove the `data` key from the argument object passed to `responseMeta` (because with a streamed response, the headers are sent before the data is available)

## Compatibility (client-side)

Browser support should be identical to [`fetch`](https://caniuse.com/fetch) support.

For runtimes other than the browser ones, the `fetch` implementation should support streaming, meaning that the response obtained by `await fetch(...)` should have a `body` property of type `ReadableStream<Uint8Array> | NodeJS.ReadableStream`, meaning that:

- either `response.body.getReader` is a function that returns a `ReadableStreamDefaultReader<Uint8Array>` object
- or `response.body` is a `Uint8Array` `Buffer`

This includes support for `undici`, `node-fetch`, native Node.js fetch implementation, and WebAPI fetch implementation (browsers).

## Compatibility (server-side)

> > ⚠️ for **aws lambda**, `unstable_httpBatchStreamLink` is not supported (will simply behave like a regular `httpBatchStreamLink`). It should not break anything if enabled, but will not have any effect.

Author: Same as above, we're actually referring to `httpBatchLink` here because for aws lambda, both links behave the same.

```suggestion
> ⚠️ for **aws lambda**, `unstable_httpBatchStreamLink` is not supported (will simply behave like a regular `httpBatchLink`). It should not break anything if enabled, but will not have any effect.
```
